build-logic/src/main/kotlin/dokkabuild.test-k2.gradle.kts

```diff
@@ -9,39 +12,82 @@
 plugins {
     id("dokkabuild.base")
     id("dokkabuild.java")
+    `jvm-test-suite`
 }
 
-val descriptorsTestConfiguration: Configuration by configurations.creating {
-    extendsFrom(configurations.testImplementation.get())
-}
-val symbolsTestConfiguration: Configuration by configurations.creating {
-    extendsFrom(configurations.testImplementation.get())
+val symbolsTestImplementation: Configuration by configurations.creating {
+    description = "Dependencies for symbols tests"
+    declarable()
 }
 
-val symbolsTest = tasks.register<Test>("symbolsTest") {
-    useJUnitPlatform {
-        excludeTags("onlyDescriptors", "onlyDescriptorsMPP")
-    }
-    classpath += symbolsTestConfiguration
-}
-// run symbols and descriptors tests
-tasks.test {
-    //enabled = false
-    useJUnitPlatform {
-        excludeTags("onlySymbols")
-    }
-    classpath += descriptorsTestConfiguration
-    dependsOn(symbolsTest)
+val symbolsTestImplementationResolver: Configuration by configurations.creating {
+    description = "Resolve dependencies for symbols tests"
+    resolvable()
+    extendsFrom(symbolsTestImplementation)
+    attributes { jvmJar(objects) }
 }
 
-val descriptorsTest = tasks.register<Test>("descriptorsTest") {
-    useJUnitPlatform {
-        excludeTags("onlySymbols")
-    }
-    classpath += descriptorsTestConfiguration
+val descriptorsTestImplementation: Configuration by configurations.creating {
+    description = "Dependencies for descriptors tests"
+    declarable()
 }
 
-tasks.check {
-    dependsOn(symbolsTest)
+val descriptorsTestImplementationResolver: Configuration by configurations.creating {
+    description = "Resolve dependencies for descriptors tests"
+    resolvable()
+    extendsFrom(descriptorsTestImplementation)
+    attributes { jvmJar(objects) }
 }
 
+@Suppress("UnstableApiUsage")
+testing {
+    suites {
+        val test by suites.getting(JvmTestSuite::class) {
+
+            // JUnit tags for descriptors and symbols are defined with annotations in test classes.
+            val descriptorTags = listOf("onlyDescriptors", "onlyDescriptorsMPP")
+            val symbolsTags = listOf("onlySymbols")
+
+            // Configure the regular 'test' target
+            val testTarget = targets.named("test") {
+                testTask.configure {
+                    description = "Runs tests (excluding descriptor and symbols tags: ${descriptorTags + symbolsTags})"
+                    useJUnitPlatform {
+                        excludeTags.addAll(descriptorTags + symbolsTags)
+                    }
+                    classpath += descriptorsTestImplementationResolver.incoming.files
```

Comment:
It is not clear to me why we have the same configuration as for `descriptorsTest` ?

Reply:
Good question. Basically, Dokka needs to analyse source code.

* For K1 source code, it uses [a descriptor-based implementation](https://github.com/Kotlin/dokka/tree/v1.9.20/dokka-subprojects/analysis-kotlin-descriptors).
* And for K2, [symbols-based implementation](https://github.com/Kotlin/dokka/tree/v1.9.20/dokka-subprojects/analysis-kotlin-symbols).

The Dokka subprojects don't use these implementations directly, they need to be provided at runtime. So, to run the tests, Dokka needs one implementation or the other. Dokka is still being converted to K2, so at the moment K1 is the default. Hence, the K1 descriptor-based analyser is used for 'regular' unit tests.

Correct me if I'm wrong please Dokka team!

Reply:
That makes sense, thanks!

Reply:
So, overall the idea is to run ALL tests with both K1 and K2 analysis. + there are some tests where behaviour differs (because f.e K1 or K2 analysis has some bugs), so for this cases we have annotations like `OnlyDescriptors`/`OnlySymbols`

And AFAIU, approach in this PR is not what we want. If im not wrong now we have 3 test tasks:
* `test` will run ALL tests excluding marked with annotation using K1
* `descriptorsTest` will run only marked for descriptors tests with K1
* `symbolsTest` will run only marked for symbols tests with K2

So I believe it's not what we want.
I don't know how flexible are test-suites but finally we want to have two separate tasks to run analysis for K1 and K2, f.e:
* `symbolsTest` will run ALL tests excluding `OnlySymbols` with K1 runtime
* `descriptorsTest` will run ALL tests excluding `OnlyDescriptors` with K2 runtime
* `test` will just run both `symbolsTest` and `descriptorsTest`, but don't run ALL tests one more time

Does it help?

Reply:
> symbolsTest will run ALL tests excluding OnlySymbols with K1 runtime
descriptorsTest will run ALL tests excluding OnlyDescriptors with K2 runtime

it seems @whyoleg  mixed up `symbolsTest` and `descriptorsTest`
The correct order is
- `descriptorsTest` will run ALL tests excluding `OnlySymbols` with K1 runtime
- `symbolsTest` will run ALL tests excluding `OnlyDescriptors` with K2 runtime

Reply:
Oops, yes, sorry, will fix the comment :(

Reply:
Thanks! I see that this PR is a change in behaviour. 

I'm still confused about the best way to proceed though, because when I look on master, `descriptorsTest` and `symbolsTest` are exactly the same

* https://github.com/Kotlin/dokka/blob/bc58001c58a8a827a31f44820634794dd452acd6/build-logic/src/main/kotlin/dokkabuild.test-k2.gradle.kts#L28-L35
* https://github.com/Kotlin/dokka/blob/bc58001c58a8a827a31f44820634794dd452acd6/build-logic/src/main/kotlin/dokkabuild.test-k2.gradle.kts#L37-L42

What I wanted to avoid is having different tasks run the same test twice, with exactly the same dependencies.

I see that `descriptorsTest` isn't added to `check` or `test`, so it won't always run. Maybe that's the only intended purpose of this task? Do you @whyoleg, @vmishenev, run `./gradlew descriptorsTest`, just to test non-symbols tests in some subprojects?

Anyway, this PR is supposed to be a drop-in replacement so I'll update it to match what's on master, so I'm just asking because it'd help if I could add some comments to make the intention clearer :)

Reply:
So, on master we have:
* `descriptorsTest` - runs tests with K1
* `symbolsTest` - runs test with K2
* `test` - configured the same as `descriptorsTest`, so runs tests with K1 + runs `symbolsTest` - so when running `test` you will be able to run tests both with K1 and K2

> I see that descriptorsTest isn't added to check or test, so it won't always run

But, you can run it from IDEA gutter - so f.e I can run tests for some class/method with K1 only, or with K2 only
<img width="560" alt="image" src="https://github.com/Kotlin/dokka/assets/50462752/74c150cf-4616-42e7-a881-7f1aae999728">

Ideally, I think, it will be fine that `test` will not run any tests by itself, but just run dependent tasks (for K1 and K2), but I'm not sure, that it's possible to make compatible with both CLI and IDE, so current approach (from master) should be a good compromise

Does it help to understand the idea better? May be Vadim has something additional to say here

Reply:
Thanks for explaining! So the idea is to provide a more descriptive name, even if `test` and `descriptorsTest` are both the same, under the hood?

> Ideally, I think, it will be fine that `test` will not run any tests by itself, but just run dependent tasks (for K1 and K2), but I'm not sure, that it's possible to make compatible with both CLI and IDE, so current approach (from master) should be a good compromise

Yeah, that makes sense. It's what I tried to do in the original PR, but JUnit and/or Gradle doesn't like having a test tasks that don't do anything (which is very important, because if a bug means no tests run then you don't want the test task to succeed!)

Anyway, I've reverted the behaviour to master, so it should work the same. 

And here's a summary of the discussed options:

| Task           | uses        | master | initial PR | test-runs-nothing proposal | test-runs-no-tags proposal |
|:---------------|:------------|:-------|:-----------|:---------------------------|:---------------------------|
| test           | descriptors | !K2    | !K1 && !K2 | _nothing_                  | !K1 && !K2                 |
| descriptorTest | descriptors | !K2    | K1         | !K2                        | !K2                        |
| symbolTest     | symbols     | !K1    | K2         | !K1                        | !K1                        |

Reply:
may be we can then add comment the reason why `test` task is configured that way? So it's just an alias to descriptor tests, to avoid future confusions?

Reply:
aside from comment, it would be good to actualise PR description

Reply:
I have a question. Will the `test` task run the symbols tests as well?


Our CI (TeamCity/GH Actions) is set to run only the `test` task, e.g. https://github.com/Kotlin/dokka/blob/ed3ab96896d0ae5b0a5733bb70dd9bf88fb5146a/.github/workflows/tests-smoke.yml#L35